src/LiveryBridge.ts

```
@@ -0,0 +1,296 @@
+
+  protected listenerMap: Map<string, (value: any) => void>;
+
+  protected messageListener?: (message: string) => void;
+
+  protected targetWindow: Window;
+
+  protected targetWindowUrl: string;
+
+  protected version: string;
+
+  constructor(
+    targetWindow: Window,
+    targetWindowUrl: string,
+    version: string,
+    messageListener?: (message: string) => void,
+  ) {
+    this.deferredMap = new Map<string, Deferred>();
+    this.listenerMap = new Map<string, (value: any) => void>();
+
+    this.targetWindow = targetWindow;
+    this.targetWindowUrl = targetWindowUrl;
+    this.version = version;
+
+    this.messageListener = messageListener;
+
+    // Start listening for messages
+    window.addEventListener('message', (e) => {
+      if (LiveryBridge.isLiveryMessage(e.data)) {
+        this.handleMessage(e.data);
+      }
+    });
+
+    // Attempt to handshake.
+    this.handshakeId = uuid();
+    this.handshakePromise = this.sendHandshake(
+      this.handshakeId,
+      this.version,
+    ).catch((error: Error) => {
+      throw error;
+    });
+  }
+
+  static isCommandMessage(object: LiveryMessage): object is CommandMessage {
+    const command = object as CommandMessage;
+    return object.type === 'command' && !!object.id && !!command.name;
+  }
+
+  static isEventMessage(
+    object: LiveryMessage,
+  ): object is EventMessage<unknown> {
+    const event = object as EventMessage<unknown>;
+    return object.type === 'event' && !!event.id && !!event.value;
+  }
+
+  static isHandshakeMessage(object: LiveryMessage): object is HandshakeMessage {
+    return (
+      object.type === 'handshake' &&
+      !!(object as HandshakeMessage).version &&
+      !!object.id
+    );
+  }
+
+  static isLiveryMessage(object: any): object is LiveryMessage {
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access -- Incoming object is any.
+    return !!object && object.isLivery === true && !!object.type;
```

Strictly speaking you should also check that `type` is a string and even one of the types that we support. Otherwise you should throw an error specifying that the type is illegal. Likewise with all of the other checks below.

---

Is returning false, when `type` is not a string enough? In that case a reject message is sent back.

---

Other type guards already check if `type` is a string. For example: `object.type === 'reject'`